src/aleph/http/server.clj

```diff
@@ -378,6 +394,21 @@
       ([_ ctx msg]
         (cond
 
+          ;; Happens when io.netty.handler.codec.http.HttpObjectAggregator is part of the pipeline.
+          (instance? FullHttpRequest msg)
+          (if (invalid-request? msg)
+            (reject-invalid-request ctx msg)
+            (let [^FullHttpRequest req msg
+                  content (.content req)
+                  ch (netty/channel ctx)
+                  s (netty/source ch)]
+              (when-not (zero? (.readableBytes content))
+                ;; Retain the content of FullHttpRequest one extra time to
+                ;; compensate for it being released together with the request.
+                (netty/put! ch s (netty/acquire content)))
```

Comment:
The comment and the approach is correct there regarding the expected behavior of `raw-ring-handler`.

Without the `HttpObjectAggregator` on the pipeline, we are receiving first the `HttpRequest` (io.netty.buffer.ByteBuf) which doesn't contain the data of the HTTP request and we initialize the `manifold.stream`. The `HttpRequest` is released later by `handle-request`.
And then we are receiving (multiple) `HttpContent` (io.netty.buffer.ByteBuf) which will be put directly on the `manifold.stream` without any ref counting management.

When using the `HttpObjectAggregator`, we need to `retain` one more time to let the `handle-request` releases the `ByteBuf` and then the client of the stream to release the `ByteBuf` too.

An alternative would be to not perform the `retain` and to provide an additional parameter to `handle-request` to not perform the `release`. But it's an optimization that might not be needed at all but I wanted to mention it.

Reply:
(Netty 5 will remove the notion of refcounting, but this is another story...)

Reply:
> An alternative would be to not perform the retain and to provide an additional parameter to handle-request to not perform the release. But it's an optimization that might not be needed at all but I wanted to mention it.

Good thought! This would be a breaking change, though... I suspect that this wouldn't be worth it as bumping the refcount should be a pretty cheap operation.